Add kube-api audit logging how-to (#906)
spectrocloud/librarium: content/docs/10-audit-logs/05-kube-api-audit-logging.md

import InfoBox from 'shared/components/InfoBox';
import PointsOfInterest from 'shared/components/common/PointOfInterest';
import Tooltip from "shared/components/ui/Tooltip";

# Enable Kubernetes API Audit Logging

Kubernetes auditing is a feature of the Kubernetes cluster management system that allows administrators to track and log events within the cluster. Administrators can review actions taken by users and applications and changes to the cluster's configuration or state. By enabling auditing, organizations and system administrators can better understand their users' actions and behaviors. The audit log answers common questions about what, where, when, and who.

You can also meet internal security control requirements by enabling audit logging. Many security controls require the following capabilities.

- ensuring administrators can trace the actions of individual users back to a specific person.

- to debug an issue where an unknown application is modifying resources

The guidance on this page is based on the upstream Kubernetes documentation and `kube-apiserver` source code. Follow the steps below to enable audit logging for the Kubernetes API server.

<br />

<WarningBox>

> Enabling audit logging increases the memory consumption of the API server because some context required for auditing is stored for each request. Memory consumption depends on the audit logging configuration.

Reviewer: Enabling audit logging causes the API server to consume more memory, as it needs to store additional context for each request to facilitate auditing.

> Enabling audit logging increases the memory consumption of the API server because some context required for auditing is stored for each request. Memory consumption depends on the audit logging configuration.

Reviewer 2: ```suggestion
Enabling audit logging causes the API server to consume more memory, as it needs to store additional context for each request to facilitate auditing. Memory consumption depends on the audit logging configuration.
```

</WarningBox>

# Prerequisites

- Access to a Kubernetes cluster node.

- Write access to the file system.

- Remote access to the Kubernetes cluster master nodes.

# Enable Auditing

> The Kubernetes API Audit policies contain rules about what events are captured and the level of detail to include.

Reviewer: The Kubernetes API Audit policies define the rules for capturing events and specifying the level of detail to include.

> The Kubernetes API Audit policies contain rules about what events are captured and the level of detail to include.

Reviewer 2: ```suggestion
The Kubernetes API Audit policies define the rules for capturing events and specifying the level of detail to include.
```

The audit policy you create will capture all requests at the *metadata* level. To learn more about the various audit levels, visit the Kubernetes API [Audit Policy](https://kubernetes.io/docs/tasks/debug/debug-cluster/audit/#audit-policy) documentation.

<br />

1. Log in to [Palette](https://console.spectrocloud.com).

> 2. Identify one of your cluster contro-plane nodes. You find a cluster node by navigating to the left **Main Menu** and selecting **Clusters**. Click on your cluster to access the details pages and click on the **Nodes** tab. The tab contains information about each pool, select a node from the **Master Pool** to view its IP address.

Reviewer: Click on your cluster to access the details pages, and click on the **Nodes** tab. The tab contains information about each pool; select a node from the **Master Pool** to view its IP address.

Reviewer 2: I like your suggestion @rahulhazra97 but we need to tell the user how to get to the cluster details page. That's why I feel the current iteration is better to stick with.

> 2. Identify one of your cluster contro-plane nodes. You find a cluster node by navigating to the left **Main Menu** and selecting **Clusters**. Click on your cluster to access the details pages and click on the **Nodes** tab. The tab contains information about each pool, select a node from the **Master Pool** to view its IP address.

Reviewer 2: ```suggestion
2. Identify one of your cluster contro-plane nodes. You can find a cluster node by navigating to the left **Main Menu** and selecting **Clusters**. Click on your cluster to access the details pages and click on the **Nodes** tab. The tab contains information about each pool, select a node from the **Master Pool** to view its IP address.
```

> 3. SSH into one of your control-plane nodes by using its IP address and the SSH key you specified during the cluster creation process. 

Reviewer: SSH into one of your control-plane nodes using its IP address and the SSH key you specified during the cluster creation process.

> 3. SSH into one of your control-plane nodes by using its IP address and the SSH key you specified during the cluster creation process. 

Reviewer 2: ```suggestion
3. SSH into one of your control-plane nodes using its IP address and the SSH key you specified during the cluster creation process.
```

4. From a control-plane node in the target cluster, issue the following command to create your audit policy file.

  <br />

  ```bash
  cat << EOF > /etc/kubernetes/audit-policy.yaml
    apiVersion: audit.k8s.io/v1
    kind: Policy
    rules:
    - level: Metadata
  EOF
  ```
  The audit log output will be written to a file located at **/var/log/kubernetes/audit/audit.log**. In production environments, you should ensure this log file is ingested by a logging and monitoring application. 

  The **/var/log/kubernetes/audit/** directory should be backed by persistent storage to ensure that any unshipped audit logs will not be lost during an unexpected outage of the node.

<br />

> 5. Next, you will update the Kubernetes API server manifest file. The manifest file is located in the **/etc/kubernetes/manifests** folder. 

Reviewer: Next, you will update the Kubernetes API server manifest file. The manifest file is in the **/etc/kubernetes/manifests** folder. 
Before you can modify the manifest file, please create a backup copy.

> Before modifying the manifest file, create a backup copy. 

Reviewer 2: ```suggestion
Before you modify the manifest file, create a backup copy.
```

  <br />

  ```shell
  cp /etc/kubernetes/manifests/kube-apiserver.yaml  /etc/kubernetes/manifests/kube-apiserver.backup 
  ```

> 6. Now that you have a backup copy of the manifest file go ahead and open up the file **/etc/kubernetes/manifests/kube-apiserver.yaml** in a text editor such as Vi or Nano.

Reviewer: Now that you have a backup copy of the manifest file open the file **/etc/kubernetes/manifests/kube-apiserver.yaml** in a text editor such as Vi or Nano.

  <br />

  ```shell
  vi /etc/kubernetes/manifests/kube-apiserver.yaml
  ```

  Append the following YAML configuration to your kube-apiserver manifest.

  <br />

  ```yaml
    volumeMounts:
      - mountPath: /etc/kubernetes/audit-policy.yaml
        name: audit
        readOnly: true
      - mountPath: /var/log/kubernetes/audit/
        name: audit-log
        readOnly: false
    volumes:
    - name: audit
      hostPath:
        path: /etc/kubernetes/audit-policy.yaml
        type: File
    - name: audit-log
      hostPath:
        path: /var/log/kubernetes/audit/
        type: DirectoryOrCreate
  ```

7. The next step is to update the Kubernetes API parameters with audit settings. 

> The Kubernetes API parameters are located toward the top of the file. The code snippet below showcases where in the file to place the parameters.

Reviewer: The top of the file contains the Kubernetes API parameters. Refer to the code snippet below to determine where to place these parameters.

> The Kubernetes API parameters are located toward the top of the file. The code snippet below showcases where in the file to place the parameters.

Reviewer 2: ```suggestion
The top of the file contains the Kubernetes API parameters. Refer to the code snippet below to determine where to place these parameters.
```

|-----------------------------------|----------|-------------------------------------------------------------------------------------------------------------------------------|
| -`-audit-log-batch-max-wait`        | duration | The amount of time to wait before force writing the batch that hadn't reached the max size. Only used in batch mode. Ex: `"5s"` |
| `--audit-log-batch-throttle-enable` |   boolean       | Whether batching throttling is enabled. Only used in batch mode.                                                              |
| `--audit-log-batch-throttle-qps`    | float    | The maximum average number of batches per second. Only used in batch mode                                                     |

To learn more about each of the Kubernetes API server flags, visit the Kubernetes API parameter [documentation page](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/).

# Validate

You can validate that audit logs are captured by navigating to the specified audit folder in the `--audit-log-path` parameter. 

1. Log in to [Palette](https://console.spectrocloud.com).

> 2. Identify one of your cluster contro-plane nodes. You find a cluster node by navigating to the left **Main Menu** and selecting **Clusters**. Click on your cluster to access the details pages and click on the **Nodes** tab. The tab contains information about each pool, select a node from the **Master Pool** to view its IP address.

Reviewer: Click on your cluster to access the details pages, and click on the **Nodes** tab. The tab contains information about each pool; select a node from the **Master Pool** to view its IP address.

> 3. SSH into one of your control-plane nodes by using its IP address and the SSH key you specified during the cluster creation process. 

Reviewer: SSH into one of your control-plane nodes using its IP address and the SSH key you specified during the cluster creation process.

> 3. SSH into one of your control-plane nodes by using its IP address and the SSH key you specified during the cluster creation process. 

Reviewer 2: ```suggestion
3. SSH into one of your control-plane nodes using its IP address and the SSH key you specified during the cluster creation process. 
```
